### PR TITLE
feat(core): added customer phone_number and email to session token response for click to pay

### PR DESCRIPTION
**Pull request by sahkal**
## Type of Change

- [x] Enhancement

## Description

Add customer phone_number and email to session token response for click to pay

## How did you test it?

Add Ctp_mastercard as a connector and send relevant details to ctp in connector metadata

```shell
curl --location 'http://localhost:8080/account/dqwdwljildqwjd/connectors' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: test_admin' \
--data '

{
    "connector_type": "authentication_processor",
    "connector_name": "ctp_mastercard",
    "connector_account_details": {
        "auth_type": "HeaderKey",
        "api_key": "API-KEY"
    },
    "test_mode": true,
    "disabled": false,
    "payment_methods_enabled": [
        {
            "payment_method": "card",
            "payment_method_types": [
                {
                    "payment_method_type": "credit",
                    "card_networks": [
                        "Visa",
                        "Mastercard"
                    ],
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                },
                {
                    "payment_method_type": "debit",
                    "card_networks": [
                        "Visa",
                        "Mastercard"
                    ],
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                }
            ]
        }
    ],
    "business_country": "US",
    "business_label": "default",
     "metadata": {
    "dpa_id": "b6e06cc6-3018-4c4c-bbf5-9fb232615090",
    "dpa_name": "TestMerchant",
    "locale": "en_AU",
    "card_brands": ["mastercard", "visa"],
    "acquirer_bin": "545301",
    "acquirer_merchant_id": "SRC3DS",
    "merchant_category_code": "0001",
    "merchant_country_code": "US"
  }
}
'
```

update business profile with 

```shell
"is_click_to_pay_enabled": true,
"authentication_product_ids": {"click_to_pay": "mca_id"}
```

Do payments create with customer details

```shell
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_myZVgK2frn2tWuTghpqBw4YmFcVOKoQ0etjGKyZCdegwHvKiGQgZOvhGsu7ShkaU' \
--data-raw '{
    "amount": 1130,
    "currency": "USD",
    "confirm": false,
    "return_url": "https://hyperswitch.io",
    "customer": {
        "email": "john@example.com",
        "id": "cus_abcdccqjciefgh",
        "name": "John Dough",
        "phone": "9123456789"
  } 
}'
```

Look for customer details in click to pay session response

```shell
curl --location 'http://localhost:8080/payments/session_tokens' \
--header 'Content-Type: application/json' \
--header 'x-feature: integ-custom' \
--header 'api-key: pk_dev_a82510aa11054218bfbae152586fb16d' \
--data '{
    "payment_id": "pay_gEidbQrgy29Hu2JfUWto",
    "client_secret": "pay_gEidbQrgy29Hu2JfUWto_secret_PfoMiqnCRZL3uCKrNDmv",
    "wallets": []
}
'
```

Response

```shell
{
    "payment_id": "pay_gEidbQrgy29Hu2JfUWto",
    "client_secret": "pay_gEidbQrgy29Hu2JfUWto_secret_PfoMiqnCRZL3uCKrNDmv",
    "session_token": [
        {
            "wallet_name": "click_to_pay",
            "dpa_id": "b6e06cc6-3018-4c4c-bbf5-9fb232615090",
            "dpa_name": "TestMerchant",
            "locale": "en_AU",
            "card_brands": [
                "mastercard",
                "visa"
            ],
            "acquirer_bin": "545301",
            "acquirer_merchant_id": "SRC3DS",
            "merchant_category_code": "0001",
            "merchant_country_code": "US",
            "transaction_amount": "11.30",
            "transaction_currency_code": "USD",
            "phone_number": "9123456789",
            "email": "john@example.com"
        }
    ]
}
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/payments.rs [hunk 1]
```diff
@@ -3451,10 +3468,45 @@ pub async fn get_session_token_for_click_to_pay(
...
+    if customer_details.email.is_some() {
+        return Ok(());
+    }
+
+    match (
+        customer_details.phone.as_ref(),
+        customer_details.phone_country_code.as_ref(),
+    ) {
+        (Some(_), Some(_)) => Ok(()),
+        (None, None) => Err(errors::ApiErrorResponse::MissingRequiredFields {
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/payments.rs:3493]: Ideally we should be mentioning in the error that, either email or phone number is required for ctp flow. So, that merchant would do the integration accordingly right?

Comment (sahkal) [reply crates/router/src/core/payments.rs:3493]: sure, will add another level of error check

File: crates/api_models/src/payments.rs [hunk 2]
```diff
@@ -6997,6 +6997,10 @@ pub struct ClickToPaySessionResponse {
     pub transaction_amount: StringMajorUnit,
     #[schema(value_type = Currency)]
     pub transaction_currency_code: common_enums::Currency,
+    pub phone_number: Option<String>,
```
Comment (SanchithHegde) [inline crates/api_models/src/payments.rs:7000]: Can we wrap this within a `Secret`?

**Review by sai-harsha-vardhan — DISMISSED** [review]

File: crates/router/src/core/payments.rs [hunk 3]
```diff
@@ -3439,6 +3440,21 @@ pub async fn get_session_token_for_click_to_pay(
         .change_context(errors::ApiErrorResponse::PreconditionFailed {
             message: "Failed to convert amount to string major unit for clickToPay".to_string(),
         })?;
+
+    let optional_customer_details = payment_intent
+        .customer_details
+        .as_ref()
+        .map(|details| {
+            serde_json::from_value::<CustomerData>(details.clone().into_inner().expose())
```
Comment (SanchithHegde) [inline crates/router/src/core/payments.rs:3448]: Nit: We can call `parse_value()` here.

**Review by sai-harsha-vardhan — DISMISSED** [review]

File: crates/router/src/core/payments.rs [hunk 4]
```diff
@@ -3441,14 +3441,19 @@ pub async fn get_session_token_for_click_to_pay(
...
         .clone()
-        .parse_value("CustomerData")
+        .map(|details| details.parse_value("CustomerData"))
+        .transpose()
         .change_context(errors::ApiErrorResponse::InternalServerError)
         .attach_printable("Error while parsing customer data from payment intent")?;
+    let customer_details =
+        optional_customer_details.ok_or(errors::ApiErrorResponse::MissingRequiredField {
+            field_name: "customer_details",
+        })?;
```
Comment (SanchithHegde) [inline crates/router/src/core/payments.rs:3454]: We can call `get_required_value()` then call `parse_value()`, `change_context()`, etc.?

**Review by SanchithHegde — APPROVED** [review]

**Pull request merged**